js/core/utils/shadow_dom.js

```diff
@@ -60,3 +60,34 @@ export function addShadowDomStyles($element) {
 
     root.adoptedStyleSheets = [constructedStyleSheet];
 }
+
+function isPositionInElementRectangle(element, x, y) {
+    const rect = element.getBoundingClientRect();
+
+    return rect && x >= rect.left && x < rect.right && y >= rect.top && y < rect.bottom;
+}
+
+export function getShadowElementsFromPoint(x, y, root) {
+    const result = [];
+    const elementQueue = [root];
+
+    while(elementQueue.length) {
+        const el = elementQueue.shift();
+
+        for(let i = 0; i < el.childNodes.length; i++) {
+            const childNode = el.childNodes[i];
+
+            // eslint-disable-next-line no-undef
+            if(childNode.nodeType === Node.ELEMENT_NODE &&
+               isPositionInElementRectangle(childNode, x, y) &&
+               // eslint-disable-next-line no-undef
+               getComputedStyle(childNode).pointerEvents !== 'none'
```

Comment:
This code will trigger relayout, which is very expensive operation. Is it really necessary for out using cases?

Reply:
unfortunately we need to check this for overlays.
I am not sure if it's needed only for overlays, but if so, we can skip just overlay elements